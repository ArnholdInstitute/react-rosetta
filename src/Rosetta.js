import React, { Children, Component } from 'react';
import { bool, object, objectOf, string } from 'prop-types';

export default class Rosetta extends Component {
    static propTypes = {
        dictionary: bool,
        locale: string,
        translations: objectOf(object),
    }

    static defaultProps = {
        dictionary: false,
        locale: null,
        translations: {},
    }

    static childContextTypes = {
        locale: string,
        translations: object,
    }

    get locale() {
        const {locale} = this.props;
        return (
            locale ||
            (typeof navigator !== 'undefined' && (navigator.languages ? navigator.languages[0] : navigator.language)) ||
            (typeof process !== 'undefined' && process.locale) ||
            'en'
        );
    }

    getChildContext() {
        return {
            locale: this.locale,
            translations: this.translations,
        };
    }

    get translations() {
        const { dictionary, translations } = this.props;

        if (dictionary) {
            return translations;
        }
        return translations[this.locale] || translations[this.locale.split('-')[0]] || {};
    }

    render() {
        return Children.only(this.props.children);
    }
}
