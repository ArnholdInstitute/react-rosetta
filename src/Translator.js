import React, { Component } from 'react';
import { func, number, object, objectOf, oneOfType, oneOf, string } from 'prop-types';

export default class Translator extends Component {
    static propTypes = {
        id: string,
        className: string,
        defaultTranslation: string,
        onClick: func,
        style: objectOf(oneOfType([number, string])),
        tag: oneOf(['a', 'b', 'button', 'div', 'em', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', '1i', 'p', 'span', 'strong']),
    }

    static defaultProps = {
        id: null,
        className: null,
        defaultTranslation: '',
        onClick: null,
        style: null,
        tag: 'span',
    }

    static contextTypes = {
        locale: string,
        translations: oneOfType([objectOf(object), objectOf(string)]),
    }

    get tagProps() {
        const { className, onClick, style } = this.props;
        return {
            className: className || undefined,
            onClick: onClick || undefined,
            style: style || undefined,
        };
    }

    get translation() {
        const { defaultTranslation, id } = this.props;
        const { translations } = this.context;
        return (id ? (translations[id] || id) : this.translateChildren()) || defaultTranslation;
    }

    translateChildren() {
        const { children } = this.props;
        let translation;

        if (children && typeof children === 'string') {
            translation = this.translateWord(children, true) || this.translateSentence(children);
        }
        return translation;
    }

    translateSentence(sentence) {
        return sentence.split(' ').map(word => this.translateWord(word)).join(' ');
    }

    translateWord(word, skipWordFallback) {
        const { locale, translations } = this.context;
        const { defaultTranslation } = this.props;
        const dictionary = translations[word] || {};

        let translation = dictionary[locale] || dictionary[locale.split('-')[0]];
        if (!translation && !defaultTranslation && !skipWordFallback) {
            translation = word;
        }
        return translation
    }

    render() {
        const Tag = this.props.tag;
        return <Tag {...this.tagProps}>{this.translation}</Tag>;
    }
}
