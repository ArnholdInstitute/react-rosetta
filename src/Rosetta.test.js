import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Rosetta from './Rosetta';
import Translator from './Translator';

test('Rosetta should accept a hardoded locale.', () => {
    const translations = {
        en: {
            foo: 'Hello'
        },
        es: {
            foo: 'Hola'
        }
    };

    const component = mount(
        <Rosetta translations={translations} locale="es">
            <Translator id="foo" />
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Rosetta should not throw an error if no translations are found for the locale.', () => {
    const translations = {
        en: {
            foo: 'Hello'
        },
        es: {
            foo: 'Hola'
        }
    };

    const component = mount(
        <Rosetta translations={translations} locale="fr">
            <Translator id="foo" />
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Translator should be able to translate by Id.', () => {
    const translations = {
        en: {
            foo: 'Hello'
        },
        es: {
            foo: 'Hola'
        }
    };

    const component = mount(
        <Rosetta translations={translations}>
            <Translator id="foo" />
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Translator should be able to translate a single word using a dictionary.', () => {
    const translations = {
        Hello: {
            en: 'Hi',
            es: 'Hola'
        }
    };

    const component = mount(
        <Rosetta translations={translations} dictionary>
            <Translator>Hello</Translator>
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Translator should be able to translate a phrase using a dictionary.', () => {
    const translations = {
        'Hello World': {
            en: 'Hi Earth',
            es: 'Hola Mundo'
        }
    };

    const component = mount(
        <Rosetta translations={translations} dictionary>
            <Translator>Hello World</Translator>
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Translator should be able to translate a sentence using a dictionary.', () => {
    const translations = {
        Hello: {
            en: 'Hi',
            es: 'Hola'
        },
        World: {
            en: 'Earth',
            es: 'Mundo'
        }
    };

    const component = mount(
        <Rosetta locale="es" translations={translations} dictionary>
            <Translator>Hello World</Translator>
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Translator should fall back to the id if no matches are found in the translations.', () => {
    const translations = {
        en: {
            foo: 'Hello'
        },
        es: {
            foo: 'Hola'
        }
    };

    const component = mount(
        <Rosetta translations={translations} locale="es">
            <Translator id="bar" />
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Translator should fall back to the original word if no matches are found in the translations (by dictionary).', () => {
    const translations = {
        Hello: {
            en: 'Hi',
            es: 'Hola'
        },
        World: {
            en: 'Earth',
            es: 'Mundo'
        }
    };

    const component = mount(
        <Rosetta translations={translations} dictionary>
            <Translator>Foo</Translator>
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Translator should fall back to the defaultTranslation if no matches are found and a default is defined.', () => {
    const translations = {
        Hello: {
            en: 'Hi',
            es: 'Hola'
        },
        World: {
            en: 'Earth',
            es: 'Mundo'
        }
    };

    const component = mount(
        <Rosetta translations={translations} dictionary>
            <Translator defaultTranslation="Bar">Foo</Translator>
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Translator should render the default translation if no children are defined and using a dictionary', () => {
    const translations = {
        Hello: {
            en: 'Hi',
            es: 'Hola'
        },
        World: {
            en: 'Earth',
            es: 'Mundo'
        }
    };

    const component = mount(
        <Rosetta translations={translations} dictionary>
            <Translator defaultTranslation="Bar" />
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test('Translator should be able to apply className, onClick, and Style.', () => {
    const translations = {
        en: {
            foo: 'bar'
        },
        es: {
            foo: 'Hola'
        }
    };

    const component = mount(
        <Rosetta translations={translations}>
            <Translator
                id="foo"
                className="foo"
                onClick={() => console.log('foo')}
                style={{background: 'red'}}
            />
        </Rosetta>
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});
