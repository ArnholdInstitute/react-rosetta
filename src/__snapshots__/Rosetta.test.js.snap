// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rosetta should accept a hardoded locale. 1`] = `
<Rosetta
  dictionary={false}
  locale="es"
  translations={
    Object {
      "en": Object {
        "foo": "Hello",
      },
      "es": Object {
        "foo": "Hola",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation=""
    id="foo"
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      Hola
    </span>
  </Translator>
</Rosetta>
`;

exports[`Rosetta should not throw an error if no translations are found for the locale. 1`] = `
<Rosetta
  dictionary={false}
  locale="fr"
  translations={
    Object {
      "en": Object {
        "foo": "Hello",
      },
      "es": Object {
        "foo": "Hola",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation=""
    id="foo"
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      foo
    </span>
  </Translator>
</Rosetta>
`;

exports[`Translator should be able to apply className, onClick, and Style. 1`] = `
<Rosetta
  dictionary={false}
  locale={null}
  translations={
    Object {
      "en": Object {
        "foo": "bar",
      },
      "es": Object {
        "foo": "Hola",
      },
    }
  }
>
  <Translator
    className="foo"
    defaultTranslation=""
    id="foo"
    onClick={[Function]}
    style={
      Object {
        "background": "red",
      }
    }
    tag="span"
  >
    <span
      className="foo"
      onClick={[Function]}
      style={
        Object {
          "background": "red",
        }
      }
    >
      bar
    </span>
  </Translator>
</Rosetta>
`;

exports[`Translator should be able to translate a phrase using a dictionary. 1`] = `
<Rosetta
  dictionary={true}
  locale={null}
  translations={
    Object {
      "Hello World": Object {
        "en": "Hi Earth",
        "es": "Hola Mundo",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation=""
    id={null}
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      Hi Earth
    </span>
  </Translator>
</Rosetta>
`;

exports[`Translator should be able to translate a sentence using a dictionary. 1`] = `
<Rosetta
  dictionary={true}
  locale="es"
  translations={
    Object {
      "Hello": Object {
        "en": "Hi",
        "es": "Hola",
      },
      "World": Object {
        "en": "Earth",
        "es": "Mundo",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation=""
    id={null}
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      Hola Mundo
    </span>
  </Translator>
</Rosetta>
`;

exports[`Translator should be able to translate a single word using a dictionary. 1`] = `
<Rosetta
  dictionary={true}
  locale={null}
  translations={
    Object {
      "Hello": Object {
        "en": "Hi",
        "es": "Hola",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation=""
    id={null}
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      Hi
    </span>
  </Translator>
</Rosetta>
`;

exports[`Translator should be able to translate by Id. 1`] = `
<Rosetta
  dictionary={false}
  locale={null}
  translations={
    Object {
      "en": Object {
        "foo": "Hello",
      },
      "es": Object {
        "foo": "Hola",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation=""
    id="foo"
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      Hello
    </span>
  </Translator>
</Rosetta>
`;

exports[`Translator should fall back to the defaultTranslation if no matches are found and a default is defined. 1`] = `
<Rosetta
  dictionary={true}
  locale={null}
  translations={
    Object {
      "Hello": Object {
        "en": "Hi",
        "es": "Hola",
      },
      "World": Object {
        "en": "Earth",
        "es": "Mundo",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation="Bar"
    id={null}
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      Bar
    </span>
  </Translator>
</Rosetta>
`;

exports[`Translator should fall back to the id if no matches are found in the translations. 1`] = `
<Rosetta
  dictionary={false}
  locale="es"
  translations={
    Object {
      "en": Object {
        "foo": "Hello",
      },
      "es": Object {
        "foo": "Hola",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation=""
    id="bar"
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      bar
    </span>
  </Translator>
</Rosetta>
`;

exports[`Translator should fall back to the original word if no matches are found in the translations (by dictionary). 1`] = `
<Rosetta
  dictionary={true}
  locale={null}
  translations={
    Object {
      "Hello": Object {
        "en": "Hi",
        "es": "Hola",
      },
      "World": Object {
        "en": "Earth",
        "es": "Mundo",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation=""
    id={null}
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      Foo
    </span>
  </Translator>
</Rosetta>
`;

exports[`Translator should render the default translation if no children are defined and using a dictionary 1`] = `
<Rosetta
  dictionary={true}
  locale={null}
  translations={
    Object {
      "Hello": Object {
        "en": "Hi",
        "es": "Hola",
      },
      "World": Object {
        "en": "Earth",
        "es": "Mundo",
      },
    }
  }
>
  <Translator
    className={null}
    defaultTranslation="Bar"
    id={null}
    onClick={null}
    style={null}
    tag="span"
  >
    <span>
      Bar
    </span>
  </Translator>
</Rosetta>
`;
